File: Top_100_Liked_Questions/33._Search_in_Rotated_Sorted_Array/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    // iterative binary search
    let l = 0;
    let r = nums.length-1;
    
    while (l <= r) {
        let m = l + Math.floor((r-l)/2);
        if (nums[m] === target) {
            return m;
        }

        if (nums[l] < nums[r]) {
            if (target < nums[m]) {
                r = m - 1;
            } else {
                l = m + 1;
            }
        } else {
            if (nums[l] <= nums[m]) {
                if (nums[l] <= target && target < nums[m]) {
                    r = m - 1;
                } else {
                    l = m + 1;
                }
            } else if (nums[m] <= nums[r]) {
                if (nums[m] < target && target <= nums[r]) {
                    l = m + 1;
                } else {
                    r = m - 1;
                }
            }
        }
    }
    
    return -1;
}


var search2 = function(nums, target) {
    // recursive binary search
    let helper = function(l, r) {
        if (l > r) {
            return -1;
        }
        let m = l + Math.floor((r-l)/2);
        if (nums[m] === target) {
            return m;
        }
        
        if (nums[l] <= nums[m]) {
            if (nums[l] <= target && target < nums[m]) {
                return helper(l, m-1);
            } else {
                return helper(m+1, r);
            }
        } else if (nums[m] <= nums[r]) {
            if (nums[m] < target && target <= nums[r]) {
                return helper(m+1, r);
            } else {
                return helper(l, m-1);
            }
        } else {
            throw "array not sorted";
        }
    }
    
    return helper(0, nums.length-1);
};

File: Top_100_Liked_Questions/33._Search_in_Rotated_Sorted_Array/solution.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length-1;

        while (left <= right) {
            int mid = left + Math.floorDiv(right - left, 2);
            if (nums[mid] == target)
                return mid;
            else if (nums[left] <= nums[mid]) {
                if (nums[left] <= target && target < nums[mid])
                    right = mid - 1;
                else
                    left = mid + 1;
            } else if (nums[mid] <= nums[right]) {
                if (nums[mid] < target && target <= nums[right])
                    left = mid + 1;
                else
                    right = mid - 1;
            }
        }

        return -1;
    }
}


class Solution2 {
    public int search(int[] nums, int target) {
        return binarySearchRotation(nums, 0, nums.length-1, target);
    }

    public int binarySearchRotation(int[] nums, int left, int right, int target) {
        if (left > right)
            return -1;

        int mid = left + Math.floorDiv(right - left, 2);
        if (target == nums[mid])
            return mid;
        else if (nums[left] <= nums[mid]) {
            if (nums[left] <= target && target < nums[mid])
                return binarySearchRotation(nums, left, mid-1, target);
            else
                return binarySearchRotation(nums, mid+1, right, target);
        } else { // nums[mid] <= nums[right]
            if (nums[mid] < target && target <= nums[right])
                return binarySearchRotation(nums, mid+1, right, target);
            else
                return binarySearchRotation(nums, left, mid-1, target);
        }
    }
}